test/pylib/resource_gather.py

```diff
@@ -0,0 +1,219 @@
+
+        if (self.cgroup_path / 'cpu.stat').exists():
+            with open(self.cgroup_path / 'cpu.stat', 'r', ) as file:
+                self._parse_cpu_stat(file, test_metrics)
+        return test_metrics
+
+    def write_metrics_to_db(self, metrics: Metric, success: bool = False) -> None:
+        metrics.success = success
+        self.sqlite_writer.write_row(metrics, METRICS_TABLE)
+
+
+    def put_process_to_cgroup(self):
+        super().put_process_to_cgroup()
+        pid = os.getpid()
+        with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+            cgroup.write(str(pid))
+
+    def cgroup_monitor(self, test_event: Event) -> Task:
+        return asyncio.create_task(self._monitor_cgroup(test_event))
+
+    async def _monitor_cgroup(self, test_event: Event) -> None:
+        """Continuously monitors CPU and memory utilization."""
+        try:
+            while not test_event.is_set():
+                with  open(self.cgroup_path / 'memory.current', 'r') as memory_current:
+                    timeline_record = CgroupMetric(
+                        test_id=self.test_id,
+                        memory=int(memory_current.read()),
+                        timestamp=datetime.now()
+                    )
+
+                    self.sqlite_writer.write_row(timeline_record, CGROUP_MEMORY_METRICS_TABLE)
+
+                # Control the frequency of updates, for example, every 2 seconds
+                await asyncio.sleep(1)
+        except asyncio.CancelledError:
+            self.logger.warning('Job was cancelled')
+
+    @staticmethod
+    def _parse_cpu_stat(file: TextIO, metrics: Metric) -> None:
+        # Map the values from cpu.state to the model. Keys are values from cpu.stat
+        stats = {'user_usec': 'user_sec', 'system_usec': 'system_sec', 'usage_usec': 'usage_sec'}
+        for line in file.readlines():
+            stat, value = line.split(' ')
+            if stat in stats.keys():
+                setattr(metrics, stats[stat], float(value) / 1_000_000)
+
+
+def get_resource_gather(is_switched_on: bool, test, tmpdir: str) -> ResourceGather:
+    if is_switched_on:
+        return ResourceGatherOn(test, tmpdir)
+    else:
+        return ResourceGatherOff(test, tmpdir)
+
+
+def setup_cgroup(is_required: bool) -> None:
+    if is_required:
+        # check where the process is executed in podman or in docker
+        is_podman = os.access("/run/.containerenv", os.F_OK)
+        is_docker = os.getpid() == 1
+        if is_podman:
+            subprocess.run(['sudo', 'chown', '-R', f"{getpass.getuser()}:{getpass.getuser()}", '/sys/fs/cgroup'],
+                           check=True)
+        elif is_docker:
+            subprocess.run(
+                [
+                    "sudo",
+                    "mount",
+                    "-o",
+                    "remount,rw",
+                    "/sys/fs/cgroup",
+                ],
+                check=True,
```

Comment:
Do the two container environments really need such different methods to give the current user permissions?

Perhaps we should revert 34e3302c017? Though the equivalent was necessary, at least with older docker.

Perhaps recent docker can remap files written as root to the original user.

Reply:
Unfortunately, yes. It is still necessary to remount the cgroup for Docker. Change that you mentioned affects podman, not docker.

Reply:
So here, look at /proc/mounts, check if the /sys/fs/cgroup entry is `ro`, and change it to rw if it is. This way, changes in the docker/podman runtime don't affect us.

Reply:
I've modified the code to check that we in a docker container and remount only for docker. So current code does not affect local run or execution in podman.

Reply:
Ok. It would be better to only check for is_cgroup_ro, not is_docker. The idea is that is docker or podman change their behavior in the future, and we build an old release, we don't have to fight with this script.

Reply:
podman not allow remounting. If podman changes something here, we will need to modify our approach. This was my main point, we cannot predict changes in docker/podman so we cannot write a code before these changes will arrive. For docker remounting seems redundant, with privileged mode it will mount RW, without it will be RO, but there will be no way to remount it with sudo.